crates/lsp/src/handlers_format.rs

```diff
@@ -41,8 +44,148 @@ pub(crate) fn document_formatting(
+    let format_range = logical_lines
+        .iter()
+        .map(|line| text_non_whitespace_range(line))
+        .reduce(|acc, new| acc.cover(new))
+        .expect("`logical_lines` is non-empty");
+
+    // We need to wrap in an `RRoot` otherwise the comments get attached too
+    // deep in the tree. See `CommentsBuilderVisitor` in biome_formatter and the
+    // `is_root` logic. Note that `node` needs to be wrapped in at least two
+    // other nodes in order to fix this problem, and here we have an `RRoot` and
+    // `RExpressionList` that do the job.
+    //
+    // Since we only format logical lines, it is fine to wrap in an expression list.
+    let Some(exprs): Option<Vec<air_r_syntax::AnyRExpression>> = logical_lines
+        .into_iter()
+        .map(|node| air_r_syntax::AnyRExpression::cast(node))
+        .collect()
+    else {
+        tracing::warn!("Can't cast to `AnyRExpression`");
+        return Ok(None);
+    };
+
+    let list = air_r_factory::r_expression_list(exprs);
+    let eof = air_r_syntax::RSyntaxToken::new_detached(RSyntaxKind::EOF, "", vec![], vec![]);
+    let root = air_r_factory::r_root(list, eof).build();
+
+    let format_info = biome_formatter::format_sub_tree(
+        root.syntax(),
+        air_r_formatter::RFormatLanguage::new(options),
+    )?;
+
+    if format_info.range().is_none() {
+        // Happens in edge cases when biome returns a `Printed::new_empty()`
+        return Ok(None);
+    };
+
+    let mut format_text = format_info.into_code();
+
+    // Remove last hard break line from our artifical expression list
+    format_text.pop();
+    let edits = to_proto::replace_range_edit(&doc.line_index, format_range, format_text)?;
+
+    Ok(Some(edits))
+}
+
+// From biome_formatter
+fn text_non_whitespace_range<E, L>(elem: &E) -> TextRange
+where
+    E: Into<SyntaxElement<L>> + Clone,
+    L: Language,
+{
+    let elem: SyntaxElement<L> = elem.clone().into();
+
+    let start = elem
+        .leading_trivia()
+        .into_iter()
+        .flat_map(|trivia| trivia.pieces())
+        .find_map(|piece| {
+            if piece.is_whitespace() || piece.is_newline() {
+                None
+            } else {
+                Some(piece.text_range().start())
+            }
+        })
+        .unwrap_or_else(|| elem.text_trimmed_range().start());
+
+    let end = elem
+        .trailing_trivia()
+        .into_iter()
+        .flat_map(|trivia| trivia.pieces().rev())
+        .find_map(|piece| {
+            if piece.is_whitespace() || piece.is_newline() {
+                None
+            } else {
+                Some(piece.text_range().end())
+            }
+        })
+        .unwrap_or_else(|| elem.text_trimmed_range().end());
+
+    TextRange::new(start, end)
+}
+
+/// Finds consecutive logical lines. Currently that's only expressions at
+/// top-level or in a braced list.
+fn find_deepest_enclosing_logical_lines(node: RSyntaxNode, range: TextRange) -> Vec<RSyntaxNode> {
```

Comment:
It seems like you can have destructive formatting if the selection spans two different `R_EXPRESSION_LIST`s?

```r
function() {
    1 + 1
}

2 + 2
```

https://github.com/user-attachments/assets/4bdce77e-b56b-460f-adbe-c5882ade492c

Reply:
Maybe one alternative implementation would be:

- Look at the `start` of the selection. Find the deepest "top level" or "braced expression" that encloses the `start` and record all top level / braced expression nodes found along the way.
- Look at the `end` of the selection. Find the deepest "top level" or "braced expression" that encloses the `end` and record all top level / braced expression nodes found along the way.
- Starting from `root`, jointly walk the recorded paths of `root -> deepest_start` and `root -> deepest_end`. If the braced expression nodes are the same, keep walking, otherwise stop walking and remember the last braced expression/top level node that was the same between them.
- Iterate the children of that braced expression/top level node, collecting logical lines that are covered by the selection. The span of these is what you'd format.

So in the above example we would format `2 + 2` but nothing in the function. "Top level" is the node that the start and end have in common, but only `2 + 2` is the child of the top level node fully enclosed by the selection

Reply:
good catch.

There's also the issue of doing partial selections in a line that wasn't working well before. This should now be fixed.